pallets/order-book/src/cache_data_layer.rs

```diff
@@ -0,0 +1,449 @@
+    MarketSide, OrderBookId, OrderPrice, OrderVolume, PriceOrders, UserLimitOrders, UserOrders,
+};
+use assets::AssetIdOf;
+use common::cache_storage::{CacheStorageDoubleMap, CacheStorageMap};
+use common::PriceVariant;
+use frame_support::ensure;
+use frame_support::sp_runtime::DispatchError;
+use sp_runtime::traits::Zero;
+
+pub struct CacheDataLayer<T: Config> {
+    limit_orders:
+        CacheStorageDoubleMap<OrderBookId<AssetIdOf<T>>, T::OrderId, LimitOrder<T>, LimitOrders<T>>,
+    bids: CacheStorageDoubleMap<
+        OrderBookId<AssetIdOf<T>>,
+        OrderPrice,
+        PriceOrders<T::OrderId, T::MaxLimitOrdersForPrice>,
+        Bids<T>,
+    >,
+    asks: CacheStorageDoubleMap<
+        OrderBookId<AssetIdOf<T>>,
+        OrderPrice,
+        PriceOrders<T::OrderId, T::MaxLimitOrdersForPrice>,
+        Asks<T>,
+    >,
+    aggregated_bids: CacheStorageMap<
+        OrderBookId<AssetIdOf<T>>,
+        MarketSide<T::MaxSidePriceCount>,
+        AggregatedBids<T>,
+    >,
+    aggregated_asks: CacheStorageMap<
+        OrderBookId<AssetIdOf<T>>,
+        MarketSide<T::MaxSidePriceCount>,
+        AggregatedAsks<T>,
+    >,
+    user_limit_orders: CacheStorageDoubleMap<
+        T::AccountId,
+        OrderBookId<AssetIdOf<T>>,
+        UserOrders<T::OrderId, T::MaxOpenedLimitOrdersPerUser>,
+        UserLimitOrders<T>,
+    >,
+}
+
+impl<T: Config> CacheDataLayer<T> {
+    pub fn new() -> Self {
+        Self {
+            limit_orders: CacheStorageDoubleMap::new(),
+            bids: CacheStorageDoubleMap::new(),
+            asks: CacheStorageDoubleMap::new(),
+            aggregated_bids: CacheStorageMap::new(),
+            aggregated_asks: CacheStorageMap::new(),
+            user_limit_orders: CacheStorageDoubleMap::new(),
+        }
+    }
+
+    pub fn commit(&mut self) {
+        self.limit_orders.commit();
+        self.bids.commit();
+        self.asks.commit();
+        self.aggregated_bids.commit();
+        self.aggregated_asks.commit();
+        self.user_limit_orders.commit();
+    }
+
+    pub fn reset(&mut self) {
+        self.limit_orders.reset();
+        self.bids.reset();
+        self.asks.reset();
+        self.aggregated_bids.reset();
+        self.aggregated_asks.reset();
+        self.user_limit_orders.reset();
+    }
+
+    fn add_to_bids(
+        &mut self,
+        order_book_id: &OrderBookId<AssetIdOf<T>>,
+        order: &LimitOrder<T>,
+    ) -> Result<(), ()> {
+        let mut bids = self
+            .bids
+            .get(order_book_id, &order.price)
+            .cloned()
+            .unwrap_or_default();
+        bids.try_push(order.id).map_err(|_| ())?;
+        self.bids.set(order_book_id, &order.price, bids);
+        Ok(())
+    }
+
+    fn remove_from_bids(
+        &mut self,
+        order_book_id: &OrderBookId<AssetIdOf<T>>,
+        order: &LimitOrder<T>,
+    ) {
+        let mut bids = self
+            .bids
+            .get(order_book_id, &order.price)
```

Comment:
With cache storage you could also implement get_mut or mutate and use it instead of clone and set

Reply:
I can't. First of all, because `DataLayer` trait must be implemented by `StorageDataLayer` as well and it cannot return mut.
Secondly, it is hard to understand is value modified or not. Cache storage has state of item that determines should we update the value in hard storage or not. If cache storage returns mu ref, there are no ways to know, is value changed or just read.